### PR TITLE
feat: enhance property search with date filters and availability check

### LINKED ISSUE #48: Implement search endpoint
# 📘 Implement `GET /api/properties/search` Endpoint for Property Search

## 🎯 Objective
Create a secure `GET /api/properties/search` endpoint that enables users to search for properties based on criteria such as location, dates, number of guests, price range, and amenities. The implementation must follow the StellarRent monorepo structure and best practices for data querying, pagination, and performance optimization.

## 🛠️ Technical Stack
- **Backend**: Node.js & Express.js
- **Database**: Supabase
- **Validation**: Zod
- **Types**: TypeScript
- **Blockchain**: Stellar (Soroban for availability checks)
- **Testing**: Jest
- **Code Quality**: ESLint/Prettier

### Query Parameters Schema
- [ ] Define the query parameters validation schema in `apps/backend/src/types/property.types.ts`:
```typescript
export const searchPropertiesSchema = z.object({
  location: z.string().min(2, "Location must be at least 2 characters").optional(),
  from: z.string().date("Invalid date format").optional(),
  to: z.string().date("Invalid date format").optional(),
  guests: z.number().int("Guests must be an integer").positive("Guests must be positive").optional(),
  priceMin: z.number().positive("Price must be positive").optional(),
  priceMax: z.number().positive("Price must be positive").optional(),
  amenities: z.string().optional(), // Comma-separated list, e.g., "wifi,pool"
  page: z.number().int("Page must be an integer").positive("Page must be positive").default(1),
  limit: z.number().int("Limit must be an integer").positive("Limit must be positive").max(50, "Limit cannot exceed 50").default(10),
});
```
### Service Layer

Update `apps/backend/src/services/property.service.ts`:
- Implement `searchProperties` function to query Supabase with dynamic filters.
- Validate date availability using the Soroban `BookingContract`.
- Handle pagination and calculate total results.
- Add error handling for database and blockchain queries.

### Controller Layer

Update `apps/backend/src/controllers/property.controller.ts`:
- Parse and validate query parameters.
- Call the `searchProperties` service function.
- Format the response with pagination metadata.
- Handle errors with appropriate status codes and messages.

### Routes

Update `apps/backend/src/routes/property.routes.ts`:
- Add the search route.
- Apply validation middleware for query parameters.
- Ensure proper error handling.

### Testing

Create `apps/backend/src/tests/property.test.ts` (extend existing file):
- Test successful search with various filters (location, dates, guests, price, amenities).
- Test pagination (different page and limit values).
- Test availability conflicts using mock Soroban contract responses.
- Test invalid query parameters (e.g., negative guests, invalid dates).
- Test database errors and edge cases.

### Acceptance Criteria

**Input Validation**
All query parameters are properly validated:
- `location`: Optional, minimum 2 characters.
- `from` and `to`: Optional, valid date format (e.g., "2025-06-01").
- `guests`: Optional, positive integer.
- `priceMin` and `priceMax`: Optional, positive numbers.
- `amenities`: Optional, comma-separated list of valid amenities.
- `page`: Default 1, positive integer.
- `limit`: Default 10, positive integer, maximum 50.

Invalid inputs return a 400 status code with detailed error messages.

**Availability Check**
- The endpoint verifies date availability using the Soroban `BookingContract` to ensure no overlaps with existing bookings.
- Properties with conflicting dates are excluded from the results.

**Response Format**

**Success (200)**
```json
{
  "properties": [
    {
      "id": "prop-123",
      "title": "Departamento en Palermo",
      "price": 100,
      "location": {
        "city": "Buenos Aires",
        "country": "Argentina"
      },
      "image": "https://supabase-url.com/image1.jpg",
      "availability": [
        {
          "from": "2025-06-01",
          "to": "2025-06-10"
        }
      ],
      "amenities": ["wifi", "cocina"]
    }
  ],
  "total": 25
}
```

**Validation Error (400)**
```json
{
  "error": "Validation error",
  "details": [
    {
      "field": "guests",
      "message": "Guests must be positive"
    }
  ]
}
```

### Error Handling

- Validation errors: 400 with detailed messages.
- Database errors: 500 with a generic message ("Internal Server Error").
- Blockchain errors (Soroban): 503 with a message ("Unable to verify availability, please try again later").

### References

- Supabase Documentation
- Express.js Routing
- Zod Schema Validation
- Soroban Documentation

### Dependencies

- Supabase connection configured.
- Soroban `BookingContract` deployed and accessible.
- Error handling middleware implemented.
- `Property` table in Supabase already created.

### Security Considerations

- Sanitize all query parameters to prevent SQL injection.
- Ensure proper CORS configuration.
- Rate limit the endpoint to prevent abuse.
- Log all search requests for monitoring and debugging.
- Validate date ranges to prevent invalid queries.

### PR DESCRIPTION
**Pull request by chiefDevOvrlrd**
<p align="center"> StellarRent Logo </p>

### 📝 Summary
-Implements blockchain-based property availability checks using the Soroban BookingContract.
-Integrates contract call logic in the backend to filter out properties with booking conflicts.
-Adds robust error handling for missing environment variables and blockchain failures.
-Ensures only properties available for the requested dates (as verified on-chain) are returned in search results.
-Improves code safety by removing non-null assertions and adding runtime checks for required config.

### 🔗 Related Issues
Closes #48 

### 🔄 Changes Made
I integrated Soroban blockchain availability checks into my property search backend, improved error handling for validation, database, and blockchain errors, updated environment variables, and addressed project setup issues including Husky and Bun. I also discussed the need to set up and seed my Supabase database for full testing.

#### ✅ Testing Checklist
- [x] Unit tests added/modified
- [ ] Integration tests performed
- [ ] Manual tests executed
- [ ] All tests pass in CI/CD

### CODE REVIEW

File: apps/backend/src/blockchain/bookingContract.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      fee: '100',
+      networkPassphrase, // or your network
+    })
+      .addOperation(
+        contract.call('check_availability', propertyIdScVal, startDateScVal, endDateScVal)
+      )
+      .setTimeout(30)
+      .build();
+
+    const sim = await server.simulateTransaction(tx);
```
[the remark this answers is not in the document]
Comment (chiefDevOvrlrd) [reply apps/backend/src/blockchain/bookingContract.ts:80]: yes i do, i wanted to even ask on the telegram but nobody answered me

**Review by respp — APPROVED** [review]
Great work on this PR — everything looks clean and well-structured, and you did the corrections really well. Regarding the contract call, that part wasn’t clearly defined from the start, so I’ll be creating more contract-related issues in the future to improve it. I’d love to invite you to keep participating and contribute again to StellarRent — thank you for contributing!
